### PR TITLE
Check that warps are only accessing the subpartition of TMem that it can access

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Warp `i` can only access to subpartition `i % 4`.

This PR is stacked on https://github.com/NVIDIA/Fuser/pull/4015

### CODE REVIEW

File: csrc/device_lower/utils.cpp
Comment (zasdfgbnm) [inline csrc/device_lower/utils.cpp:1]: This is updated to handle an edge case

**Comment by zasdfgbnm** [comment]
!test

**Comment by zasdfgbnm** [comment]
!test

**Comment by zasdfgbnm** [comment]
!test

**Comment by zasdfgbnm** [comment]
!test

File: csrc/device_lower/utils.cpp [hunk 1]
```diff
@@ -1483,8 +1483,6 @@ Projection propagate(
     const ExprGroup& eg,
     Direction direction) {
   // Just recursively propagate subtree.
-  auto from = fromGroups(id_graph, eg, direction);
-  auto to = toGroups(id_graph, eg, direction);
```
Comment (zasdfgbnm) [inline csrc/device_lower/utils.cpp:1487]: Unused variable. No idea why clang-tidy was not complaining.

**Comment by zasdfgbnm** [comment]
!test

File: csrc/device_lower/analysis/tensor_memory.cpp [hunk 2]
```diff
@@ -482,7 +482,28 @@ computeTMemLdStDataPath(Fusion* fusion, const TMemAlllocationInfo& allocation) {
       }
       NVF_THROW(error.str());
     }
-    // TODO: Validate that we are accessing the correct sub-partition
+    // Validate that warps are accessing the correct sub-partition
+    AbstractTensor t = pdims;
+    t.split(-1, 32);
+    t.split(-2, 4);
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/device_lower/analysis/tensor_memory.cpp:489]: Added a comment above:
```C++
// Warp i can only access the sub-partition i % 4
```

[the remark this answers is not in the document]
Comment (rdspring1) [reply csrc/device_lower/analysis/tensor_memory.cpp:489]: `Warp i can only access the sub-partition i % 4.` sounds more informative than `Warps are not accessing the correct sub-partition.` in the error message.

**Review by rdspring1 — COMMENTED** [review]
LGTM

File: csrc/device_lower/analysis/tensor_memory.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/device_lower/analysis/tensor_memory.cpp:489]: Thanks. Does the linearity guarantee `t.split(-2, 4)` is divisible? Or, doesn't it need to be divisible?

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/device_lower/analysis/tensor_memory.cpp:489]: No, neither `proveLinearAndGetStrideAfterPropagation` nor the hardware requires it to be divisible. It is totally OK if your kernel only has one warp, and in this case, you only have access to the first subpartition of TMem.

**Review by naoyam — APPROVED** [review]
LGTM

**Comment by zasdfgbnm** [comment]
!build
